syntax = "proto3";
package protocol;

import "common.proto";

//PBFT protocol
message PbftPrePrepare
{
	int64 view_number = 1;
	int64 sequence = 2;
	int64 replica_id = 3;
	bytes value = 4;
	bytes value_digest = 5;
}

message PbftPrepare
{
	int64 view_number = 1;
	int64 sequence = 2;
	int64 replica_id = 3;
	bytes value_digest = 4;
}

message PbftCommit
{
	int64 view_number = 1;
	int64 sequence = 2;
	int64 replica_id = 3;
	bytes value_digest = 4;
}

message PbftPreparedSet
{
	PbftEnv pre_prepare = 1;
	repeated PbftEnv prepare = 2;
}

message PbftViewChange
{
	int64 view_number = 1; //v+1
	int64 sequence = 2;     //n
	repeated PbftPreparedSet prepared_set = 3;  //prepared messages large than n
	int64 replica_id = 4;
}

message PbftNewView
{
	int64 view_number = 1; //v+1
	int64 sequence = 2;     //sequence
	int64 replica_id = 3;
	repeated PbftEnv view_changes = 4;     //V
	repeated PbftEnv pre_prepares = 5;  //O
}

enum PbftMessageType {
	  PBFT_TYPE_PREPREPARE = 0;
	  PBFT_TYPE_PREPARE = 1;
	  PBFT_TYPE_COMMIT = 2;
	  PBFT_TYPE_VIEWCHANGE = 3;
	  PBFT_TYPE_NEWVIEW = 4;
}

enum PbftValueType {
	    PBFT_VALUE_TX = 0;
	    PBFT_VALUE_TXSET = 1;
}

message Pbft
{
	int64 round_number = 1; //for send again
	PbftMessageType type = 2;
	PbftPrePrepare pre_prepare = 3;
	PbftPrepare prepare = 4;
	PbftCommit commit = 5;
	PbftViewChange view_change = 6;
	PbftNewView new_view = 7;
}

message PbftEnv
{
 Pbft pbft = 1;
 Signature signature = 2;
}

message Validator{
	string address = 1;
	int64 pledge_coin_amount = 2;
}

message ValidatorSet{
	repeated Validator validators = 1;
}

message PbftProof
{
	repeated PbftEnv commits = 1;
}

message FeeConfig
{
	enum Type {
			UNKNOWN = 0;
			GAS_PRICE					= 1;
			BASE_RESERVE				= 2;
	};
	int64 gas_price = 1;
	int64 base_reserve = 2;
}